app/workers/user_cleaner.rb

```diff
@@ -0,0 +1,121 @@
+# Deletes inactive users from the database.
+# See [1] for a description of how the flow works on a high level.
+#
+# Users have a deletion_date field that is nil by default. It is set to a future
+# date if the user has been inactive for too long (i.e. hasn't logged in).
+# Before the deletion date is reached, we send warning mails. If users log in
+# before the deletion date, that date is reset to nil such that the user is not
+# deleted. If the user is still inactive on the deletion date, the user is
+# ultimately deleted.
+#
+# [1] https://github.com/MaMpf-HD/mampf/issues/410#issuecomment-2136875776
+class UserCleaner
+  # The maximum number of users that can be deleted in one run.
+  # This is equivalent to the maximum of number of deletion dates set in one run.
+  #
+  # This flag can be used to prevent too many mails from being sent at once.
+  # Keep in mind that the mail server also handles other mails, e.g. notification
+  # mails etc., so we might want to set the limit very low here such that our
+  # mail server is not marked as "spam server".
+  #
+  # Note that this is just a soft limit, i.e. the actual number of deletion
+  # warning mails sent on a given day might be higher than this number:
+  # - If on a given day the cronjob is not run (for whatever reason),
+  #   we have more users with a deletion date lying in the past than
+  #   MAX_DELETIONS_PER_RUN. However, we don't send an additional mail once
+  #   the user is deleted, so this shouldn't be a problem.
+  # - Despite that there cannot be more than MAX_DELETIONS_PER_RUN users with the
+  #   same deletion date, warning mails might be sent on the same date to users
+  #   with varying deletion dates, since the 40-, 14-, 7- and 2-day warning mails
+  #   can overlap temporally.
+  MAX_DELETIONS_PER_RUN = 50
+
+  # Returns all users who have been inactive for 6 months,
+  # i.e. their last sign-in date is more than 6 months ago.
+  def inactive_users
+    User.where("last_sign_in_at < ?", 6.months.ago)
+  end
+
+  # Sets the deletion date for inactive users and sends an initial warning mail.
+  #
+  # This method finds all inactive users whose deletion date is nil (not set yet)
+  # and updates their deletion date to be 40 days from the current date.
+  #
+  # The maximum number of deletion dates set in one run is limited by
+  # MAX_DELETIONS_PER_RUN.
+  def set_deletion_date_for_inactive_users
+    inactive_users.where(deletion_date: nil)
+                  .limit(MAX_DELETIONS_PER_RUN)
+                  .find_each do |user|
+      user.update(deletion_date: Date.current + 40.days)
+      UserCleanerMailer.with(user: user).pending_deletion_email(40).deliver_later
```

Comment:
We send out pending deletion mails also to non-generic users even if their accounts will not be deleted in the end. This might be confusing to the recipients of the mails as they might think that their account was deleted when it actually was not.

Reply:
Ups, that's true. Fixed in 28788595